docs_input/basics/sparse_tensor.rst

```diff
@@ -0,0 +1,257 @@
+flexibility in changing storage formats by merely changing annotations in the
+type definitions, which allows for rapid experimentation with different ways
+of storing sparse tensors in a MatX computation.
+
+Quick Start
+-----------
+
+Despite the forward looking design of using the UST type, the current
+experimental support provides a few factory methods with the common
+formats COO, CSR, and CSC. The factory methods look similar to e.g.
+sparse construction methods found in SciPy sparse or torch sparse.
+
+For example, to create a COO representation of the following
+4x8 matrix with 5 nonzero elements::
+
+       | 1, 2, 0, 0, 0, 0, 0, 0 |
+   A = | 0, 0, 0, 0, 0, 0, 0, 0 |
+       | 0, 0, 0, 0, 0, 0, 0, 0 |
+       | 0, 0, 3, 4, 0, 5, 0, 0 |
+
+First, using a uniform memory space, set up the constituent 1-dim buffers
+that contain, respectively, the value, i-index, and j-index of each nonzero
+element, ordered lexicographically by row-then-column index, as follows::
+  
+  auto vals = make_tensor<float>({5});
+  auto idxi = make_tensor<int>({5});
+  auto idxj = make_tensor<int>({5});
+  vals.SetVals({1, 2, 3, 4, 5});
+  idxi.SetVals({0, 0, 3, 3, 3});
+  idxj.SetVals({0, 1, 2, 3, 5});
+
+Then, the COO representation of ``A``, residing in the same memory space as
+its constituent buffers is constructed as follows::
+
+  auto Acoo = experimental::make_tensor_coo(vals, idxi, idxj, {4, 8});
+
+  print(Acoo);
+
+The result of the print statement is shown below::
+
+  tensor_impl_2_f32: SparseTensor{float} Rank: 2, Sizes:[4, 8], Levels:[4, 8]
+  nse    = 5
+  format = ( d0, d1 ) -> ( d0 : compressed(non-unique), d1 : singleton )
+  pos[0] = ( 0  5 )
+  crd[0] = ( 0  0  3  3  3 )
+  crd[1] = ( 0  1  2  3  5 )
+  values = ( 1.0000e+00  2.0000e+00  3.0000e+00  4.0000e+00  5.0000e+00 )
+  space  = CUDA managed memory
+
+Note that, like dense tensors, sparse tensors provide ()-operations
+for indexing.  However, users should **never** use the ()-operator
+in performance critical code, since sparse storage formats do not
+provide O(1) random access to their elements (compressed levels will
+use some form of search to determine if an element is present)::
+
+  // Naive way to convert the sparse matrix back to a dense matrix.
+  auto A = make_tensor<float>({4, 8});
+  for (index_t i = 0; i < 4; i++) {
+    for (index_t j = 0; j < 8; j++) {
+      A(i, j) = Acoo(i, j);
+    }
+  }
+
+Instead, conversions (and other operations) should use sparse operations
+that are specifically optimized for the sparse storage format. The
+correct way of performing the conversion above is as follows::
+
+  auto A = make_tensor<float>({4, 8});
+  (A = sparse2dense(Acoo)).run(exec);
+
+The current experimental sparse support in MatX provides efficient
+operations for sparse-to-dense, dense-to-sparse, matmul, and solve::
+
+   (A = sparse2dense(Acoo)).run(exec);
+   (Acoo = dense2sparse(D)).run(exec);
+   (C = matmul(Acoo, B)).run(exec);
```

Comment:
Eventually update matmul.rst and make a new solve.rst for these

Reply:
ack